### DISCUSSION TITLE
Dragging nodes with the mouse

### DISCUSSION THREAD

**Discussion by MILKMILKKING** [vslinko/obsidian-outliner#190] [Archive]
Dragging nodes with the mouse, currently it seems that you can only move nodes up and down by shortcut keys, it would be much easier if you could use the mouse to drag nodes in preview mode!

**Comment by dcoales** [comment]
I would also love this but would mostly want to be able to drag them in edit mode (I pretty much live in edit mode).  

In an ideal world I could just drag a single bullet OR highlight a set of lines and then drag all selected lines (similar to how I can highlight bunch of lines and then use a single command to delete them - the highlight doesn't have to cover the whole line).

CodeMirror does appear to support drag and drop so hopefully something like this would be possible :-)

**Comment by vslinko** [comment]
@MILKMILKKING Hi, thanks for the idea! I think it should be implemented at some point.

**Reply by MILKMILKKING** [reply]
Thanks for all the hard work.Hopefully it will be possible to drag and drop nodes with the mouse in both edit mode and preview mode.

**Comment by exceptionraiser** [comment]
@vslinko Any update on this becoming a feature?

**Comment by joelostblom** [comment]
I think this behavior is especially useful on mobile where rearranging lists quickly is tricky via shortcuts. Google Keep provides this in a way that makes it one of the smoothest ways or working with lists on my phone. Would love if there was an obsidian plugin that provided this functionality.

**Reply by codey-badger** [reply]
@joelostblom hi, did you find a workaround? I was looking for something that gets close to Google keep but to no avail.

**Comment by dmkravets** [comment]
I'd be stoked with a drag and drop feature baked into the outliner tool. 
There is this plugin: https://github.com/artem-barmin/obsidian-block-drag-n-drop
It's too buggy to use.
Perhaps the devs of Outliner and this block-drag-n-drop plugin can join forces to build out this feature!

**Comment by danrthompson** [comment]
I would **love** this feature. It’s a small thing but the way this works effortlessly in roam has spoiled me. And even though I’ve fully migrated to obsidian and like it more, I still miss this feature.

I’ve tried the drag and drop blocks plugin but it is currently broken and even before it was fairly buggy.

Given how popular this extension is, and given how many people would like this feature (I think  many people who use outliner apps), you guys should add this feature. I use keyboard shortcuts very very extensively, but still it’s often easier to drag and drop something.


I totally understand if you’re too busy though and thanks very very much for making an amazing plugin.

Also, I have made a few obsidian plug-ins for myself, so I have a tiny bit of experience here although I should emphasize that it isn’t much and I’ve never built a feature like this in obsidian. But for what little it’s worth, I’d be happy to help out with this feature if there’s anything I can do.

Thanks again,
Dan

PS (I can open a separate discussion page or issue for this if you want)

As a very small aside, something that would make manuouting outline elements with the keyboard would be that if you copy or cut a line by copying with nothing selected (this part works), you should be able to just put your cursor on a line and hit paste and have it insert the line above the line your cursor is on. The current behavior is to paste the copied code into the line, where the cursor is. I get that that often, usually, is the better answer, but at minimum when a line is copied using the copy with nothing selected method, it would be nice for paste to behave accordingly.

**Comment by vslinko** [comment]
Trying to implement dnd https://github.com/vslinko/obsidian-outliner/pull/426

**Comment by vslinko** [comment]
Drag-and-drop feature is up and running!

Please update the plugin to [version 4.5.0](https://github.com/vslinko/obsidian-outliner/releases/tag/4.5.0) and enable `Drag-and-Drop (Experimental)` in the settings.

Looking forward to your feedback!

**Comment by danrthompson** [comment]
You guys are **awesome** for implementing this!! I actually requested this feature for this plugin in the forum within the last month!

Thanks so much, from the heart ❤️ . If I can find a donate link, I'll chip in.

Now just a few pieces of feedback since you asked. I want to put it in context that this feature is **awesome** and works wonderfully. But this is the feedback that I would give if this were a dedicated outliner app
- It would be nice to be able to drag items from one list to another
- I apologize in advance for the length of the following comment. If it makes sense to you, great. If not, I can include a video to show what I mean: When I've grabbed an item, sometimes it's not intuitive what I should do with my mouse to determine where it should go. Vertically is fine, but the indentation levels are sometimes a little confusing. I've noticed that sometimes the indentation level changes when I move my mouse vertically. Other times it changes when I move it horizontally. And the horizontal "threshold" of where my mouse needs to be to trigger indentation seems to vary based on where my mouse was when it was moving vertically.

**Reply by vslinko** [reply]
Thank you for the feedback!

> I apologize in advance for the length of the following comment. If it makes sense to you, great. If not, I can include a video to show what I mean: When I've grabbed an item, sometimes it's not intuitive what I should do with my mouse to determine where it should go. Vertically is fine, but the indentation levels are sometimes a little confusing. I've noticed that sometimes the indentation level changes when I move my mouse vertically. Other times it changes when I move it horizontally. And the horizontal "threshold" of where my mouse needs to be to trigger indentation seems to vary based on where my mouse was when it was moving vertically.

Yes, I agree, it can be confusing. But I have no good ideas how to do it better with all the constraints. Could you please open a new discussion on this to be sure it's not lost.

**Reply by danrthompson** [reply]
@vslinko I think possibly a good solution would be to look at Todoist. Their drag and drop functionality is very well implemented. I don't know what the rules are with Todoist, but it could help you come up with a better UX.

I think one key thing is that the indentation level should be based on whether the dragged cursor is to the left or the right of the leftmost element in the list item you're dropping the task under.

**Reply by vslinko** [reply]
Unfortunately, it is difficult to make it like in Todois because of the CodeMirror constraints. But I made some UX improvements. Please check it out https://github.com/vslinko/obsidian-outliner/releases/tag/4.6.7

**Reply by danrthompson** [reply]
Thanks so much for adding this. The improvements have helped. But honestly, it was just a wonderful feature to start with and SO useful, so these are such minor inconveniences compared to finally being able to drag and drop!!! I have been waiting for **someone** to add this for so long.

So I just wanted to say again - improvements are always welcome, but just adding this feature at all is the most important thing.

I do have a couple thoughts about possible ways to extend this awesome feature. I'm sure you've already thought of them because they were a part of the drag-and-drop-blocks plugin (which I'm sure you took inspiration from), and also they match with the Roam workflow, which you're likely familiar with.

So because of that, I'll just mention them here. But if you think it would be worth adding these as separate feature request issues I can do that. If you already have considered these features and don't need me to add them, that's fine too.

===

Basically it would be great, for me, if you supported different behavior for different types of dragging and dropping.

1. This is the least likely to make sense because it's outside your plugin's purview, but it would also be the most **incredible** feature: Adding support for dragging and dropping outside of outlines (as an optional feature). If you really feel like that is outside of the purview of the Outliner extension, even as an opt-in setting, then maybe you could create a simple separate plugin that submodules whatever code you used to implement the UI of the drag and drop feature.

I know this one sounds a bit crazy, **but** you have already built this amazing drag-and-drop functionality. And I believe this is the only Obsidian plugin to have done that. At least I couldn't find any after drag and drop blocks stopped working. So if it wouldn't be a huge dev lift, it could be a way to add functionality that I assume a **lot** of users would find really useful: being able to reorder paragraphs and headings as if they were outline items.

I do understand that this is a reach ask though. The rest of the ideas are more within Outliner's domain:

2. Most important: make it possible to drag between two different lists.

3. Add the possibility to do different types of dragging, a la Roam. Rather than moving an item, you drag and drop it to copy the item or create a link to the item in the new location. You use different modifiers to specify which action to take (move item, create link to item, create embedding to item, copy item).

I know I've already mentioned this, but making at least the non-moving functionality possible for regular blocks as well as outline items would be incredible, and it would likely not be a huge additional amount of effort when you're not moving, or dealing with indentation, but rather just copying a link to a particular line. And if you are doing the copying **from or to a list** then you really have a leg up in terms of implementing this. But again, it's OK if this is not a priority for you.

But basically I'm talking about being able to:
a. Option drag, control drag, shift drag, and/or command drag, each with different functionality
b. The different functionality options that I know of are:
- Dragging an item (list item, or maybe even text block) to another location, but rather than moving the item to that spot, it copies the item
- Dragging an item and inserting a link to the item in the new spot instead of moving it there (with a reasonable link title)
- Dragging an item and inserting an embed to the item in the new spot
c. This should **definitely** work __across__ tabs. So to start with, you should be able to drag an item from one list to another (and perhaps from a list to a random location or vice versa) within a file. Even across lists. But where these non-moving actions really shine is when you can do it across files. Basically it makes it incredibly easy to create a block or heading or list item block level reference, or copy information, or embed it, from one file to another. It's one of the core features of Roam, it is widely beloved, and I am so surprised it hasn't been implemented in Obsidian. But maybe you guys could do that!

OK sorry for the extremely long post. I'm probably procrastinating a bit. But this would be awesome. So let me know if you want me to break this out into its own issue, or possibly even multiple issues. Or if you already have considered this approach and decided for or against it, so it doesn't matter.

**Reply by vslinko** [reply]
@danrthompson thanks again! could you please create a separate feature request? don't want to lose it.

**Comment by scchearn** [comment]
This plugin is fantastic! It simplifies the process of working with bullet point lists immensely. It would be great if it could also be used for checklists. Currently, I'm only able to drag and relocate a checkbox if it has sub-items, not a standalone checkbox.

**Reply by vslinko** [reply]
Thank you for the feedback!

> It would be great if it could also be used for checklists

Released in https://github.com/vslinko/obsidian-outliner/releases/tag/4.6.4

**Comment by athanhat** [comment]
Hi everyone, excellent plugin, a useful note here for newcomers, the DnD feature works only when you activate the source mode, my setup: 
Obsidian 1.4.14 on Linux
Outliner 4.8.0
Hope that helps others

**Reply by TomSpoct** [reply]
Works but only on certain headers. Its seems that after the third header of level 1 headers does not work.
Drag and drop buggy.
Also cannot drag and drop bullets from one header but only within one header

**Comment by charlie17** [comment]
DnD does not seem to work for me. I see "the hand" when hovering over the bullet, but when clicking and attempting to drag, nothing haooens. It doesn't actually move anything. What am I missing?

**Reply by thorstenkoester** [reply]
Same here. Drag and drop is working in source mode only. Keyboard shortcuts stopped working too.

**Reply by EightWheels** [reply]
Yep, DND works only in source mode for me -- most of the time. Twice it worked in live edit mode, but I can't repeat it either in the same note or a new note. But the keyboard shortcuts have been working for me. This would be perfect, though, if it worked in live edit mode!

**Comment by charlie17** [comment]
I have discovered something by accident - DnD seems to work properly when all outline nodes in the file are expanded. As soon as some are collapsed it stops working
